src/Servers/Kestrel/test/InMemory.FunctionalTests/TlsListenerTests.cs

```diff
@@ -66,4 +68,98 @@ await sslStream.AuthenticateAsClientAsync(new SslClientAuthenticationOptions
 
         Assert.True(tlsClientHelloCallbackInvoked);
     }
+
+    [Fact]
+    public async Task TlsClientHelloBytesCallback_PreCanceledToken()
+    {
+        var tlsClientHelloCallbackInvoked = false;
+
+        var testContext = new TestServiceContext(LoggerFactory);
+        await using (var server = new TestServer(context => Task.CompletedTask,
+            testContext,
+            listenOptions =>
+            {
+                listenOptions.UseHttps(_x509Certificate2, httpsOptions =>
+                {
+                    httpsOptions.TlsClientHelloBytesCallback = (connection, clientHelloBytes) =>
+                    {
+                        Logger.LogDebug("[Received TlsClientHelloBytesCallback] Connection: {0}; TLS client hello buffer: {1}", connection.ConnectionId, clientHelloBytes.Length);
+                        tlsClientHelloCallbackInvoked = true;
+                        Assert.True(clientHelloBytes.Length > 32);
+                        Assert.NotNull(connection);
+                    };
+                });
+            }))
+        {
+            using (var connection = server.CreateConnection())
+            {
+                using (var sslStream = new SslStream(connection.Stream, false, (sender, cert, chain, errors) => true, null))
```

Comment:
I don't think these tests should actually send any bytes over the connection. That creates a race where the timeout doesn't actually occur by the time the tls client hello is received and parsed.

Reply:
HttpsConnectionMiddleware does catch the exception, so i am only seeing if the connection is closed after the timeout. I changed to 
```
await connection.TransportConnection.Input.WriteAsync(new byte[] { 0x16 });
var readResult = await connection.TransportConnection.Output.ReadAsync();

// HttpsConnectionMiddleware catches the exception, so we can only check the effects of the timeout here
Assert.True(readResult.IsCompleted);
```